Multi-Sequence LLM prompt processing and cost reductions

sebhaan opened the pull request:
Main changes:
- added and tested multi-sequence prompt processing in module experiment_multi.py
- added change from single sequence pair per prompt to multi-sequence pairs per prompt in llmprocess.py. Default currently at 8 input sequence pairs per prompt (haven't tested larger numbers yet). This significantly reduces API costs while keeping same accuracy.
- improved json formatting for LLM response. Default response file is now also saved as json file.
- changed llm default format from completion to chatcompletion to enable new gpt-3.5-turbo model model.
- removed in result table 'prompt_id' and  'confidence' columns since 'confidence' not straightforward anymore with multi sequence output, and  'prompt_id' not needed.
- added estimate_compute_cost function to llmprocess.py to estimate compute costs and time (need avg time and tokens as argument). The current costs for all OpenAI models are compiled in a new json file in schemas/openai_pricing.json.
- changed order of examples and response format to reasoning first, then classification, which should improve response accuracy.
- Added new default LLM instruction file:  'instruction_multiprompt.txt'.
- Updated test function for llmprocess.py

sebhaan committed:
Add API pricing list

sebhaan committed:
Add cost and time estimate function

sebhaan committed:
Add multiprompt instructions

sebhaan committed:
Add multiprompt generation

sebhaan committed:
Change to reasoning first

sebhaan committed:
Change llm results to multi sequences

sebhaan committed:
Udpate json result processing

sebhaan committed:
Fix json parsing and add time buffer for API

sebhaan committed:
Add NA to eval_exp

sebhaan committed:
Chnage order of reasoning

sebhaan committed:
Chunk prompt samples

sebhaan committed:
Change output parsing to multiprompt

sebhaan committed:
Add multi sequencing instructions

sebhaan committed:
Update price estimate

sebhaan committed:
Change default path directories to package directory

sebhaan committed:
Fix nseq

sebhaan committed:
Fix cost estimate and add compute test time

sebhaan committed:
Update with multi seq prompt

h-croser commented:
In llmprocess.py, if the conditional on line 475 is False, then `list_reasoning`, `list_class_pred`, and `list_linkage_pred` aren't assigned but are referenced later

Commit:
In annotation module:
- changed default instruction file path
- removed confidence column from expected LLM results

h-croser commented:
Commit [3ff23a2](https://github.com/Sydney-Informatics-Hub/LLM-LCT-sequencing/pull/2/commits/3ff23a247c44a3ff53cf6ea0629f82ceb7ea7f32) makes two minor changes to annotation module to conform to the removal of 'confidence' column and change of default prompt path

sebhaan committed:
Add NONE values to variables if conditional on line 475 is False

sebhaan commented:
> In llmprocess.py, if the conditional on line 475 is False, then `list_reasoning`, `list_class_pred`, and `list_linkage_pred` aren't assigned but are referenced later

Thanks, fixed it in commit 0b265f84bc62b5226a7dc0382a3490ad954a86db. Also changed the output values for these three columns to 'NONE' if LLM output parsing fails. Before it was 'NA' which has a different meaning, namely if the LLM can't find any sequence class.
